internal/app/singularity/pull.go

```diff
@@ -177,71 +170,66 @@ func downloadImageCallback(totalSize int64, r io.Reader, w io.Writer) error {
 func OrasPull(name, ref string, force bool, ociAuth *ocitypes.DockerAuthConfig) error {
-	ref = strings.TrimPrefix(ref, "//")
-
-	spec, err := reference.Parse(ref)
-	if err != nil {
-		return fmt.Errorf("unable to parse oci reference: %s", err)
+	if !force {
+		if _, err := os.Stat(name); err == nil {
+			return fmt.Errorf("image file already exists: %q - will not overwrite", name)
+		}
 	}
 
-	// append default tag if no object exists
-	if spec.Object == "" {
-		spec.Object = SifDefaultTag
-		sylog.Infof("No tag or digest found, using default: %s", SifDefaultTag)
+	sysCtx := &ocitypes.SystemContext{
+		DockerAuthConfig: ociAuth,
 	}
 
-	credFn := func(_ string) (string, string, error) {
-		if ociAuth != nil {
-			return ociAuth.Username, ociAuth.Password, nil
-		}
-
-		return "", "", nil
+	sum, err := oras.ImageSHA(ref, sysCtx)
+	if err != nil {
+		return fmt.Errorf("failed to get checksum for %s: %s", ref, err)
 	}
-	resolver := docker.NewResolver(docker.ResolverOptions{Credentials: credFn})
 
-	wd, err := os.Getwd()
+	imageName := uri.GetName("oras:" + ref)
+
+	imagePath := cache.OrasImage(sum, imageName)
+	exists, err := cache.OrasImageExists(sum, imageName)
 	if err != nil {
-		return fmt.Errorf("failed to get working directory: %s", err)
+		return fmt.Errorf("unable to check if %s exists: %v", imagePath, err)
 	}
 
-	store := content.NewFileStore(wd)
-	defer store.Close()
+	if !exists {
+		sylog.Infof("Downloading image with ORAS")
 
-	store.AllowPathTraversalOnWrite = true
-	store.DisableOverwrite = !force
+		if err := oras.DownloadImage(imagePath, ref, ociAuth); err != nil {
+			return fmt.Errorf("unable to Download Image: %v", err)
+		}
 
-	allowedMediaTypes := oras.WithAllowedMediaTypes([]string{SifLayerMediaType})
-	handlerFunc := func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
-		if desc.MediaType == SifLayerMediaType {
-			// Ensure descriptor is of a single file
-			// AnnotationUnpack indicates that the descriptor is of a directory
-			if desc.Annotations[content.AnnotationUnpack] == "true" {
-				return nil, fmt.Errorf("descriptor is of a bundled directory, not a SIF image")
-			}
-			nameOld, _ := content.ResolveName(desc)
-			_ = store.MapPath(nameOld, name)
+		if cacheFileHash, err := oras.ImageHash(imagePath); err != nil {
+			return fmt.Errorf("error getting ImageHash: %v", err)
+		} else if cacheFileHash != sum {
+			return fmt.Errorf("cached file hash(%s) and expected hash(%s) does not match", cacheFileHash, sum)
 		}
-		return nil, nil
+	} else {
+		sylog.Infof("Using cached image")
 	}
-	pullHandler := oras.WithPullBaseHandler(images.HandlerFunc(handlerFunc))
 
-	_, _, err = oras.Pull(orasctx.Background(), resolver, spec.String(), store, allowedMediaTypes, pullHandler)
+	// Perms are 777 *prior* to umask in order to allow image to be
+	// executed with its leading shebang like a script
+	dstFile, err := os.OpenFile(name, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0777)
```

Comment:
To prevent against a race, if force == false, you should add O_EXCL to the flags here. In that way, if the file didn't exist when you called os.Stat above, but now the file does exist, this will fail, honoring the missing --force.